app/assets/javascripts/host_edit_interfaces.js

```diff
@@ -388,13 +388,17 @@ $(document).on('change', '.virtual', function() {
     .toggle(!is_virtual);
 });
 
-function update_fqdn() {
+function construct_fqdn() {
   var host_name = $('#host_name').val();
   var domain_name = primary_nic_form()
     .find('.interface_domain option:selected')
     .text();
+  return fqdn(host_name, domain_name);
+}
```

Comment:
I think unmanaged hosts don't have NICs, but I can't check right now since I don't have an unmanaged host. Does this work in that case?

Reply:
I'll test that out, tho I've also remembered the setting `append_domain_name_for_hosts` and for that `= true` it would not work almost for sure.
I'll test both cases and see if I can figure something better.

Reply:
I do think the ID of the host is much more reliable than the FQDN, which as we see here is unstable.

Reply:
I'd agree, but you do not have the id at the time we now calculate the redirect.
I think we need a better approach. @amirfefer I've added test cases for both failing scenarios.
My approach is fixing the first, but as expected it fails for the second (`append_domain_name_for_hosts = false`)

Reply:
Shouldn't the interaction be to `POST` and then wait for the response. That should give you the correct URL. Guessing it client side is IMHO a bad way. What if the request fails? Then it will result in a 404.

Reply:
> Shouldn't the interaction be to `POST` and then wait for the response. That should give you the correct URL. Guessing it client side is IMHO a bad way. What if the request fails? Then it will result in a 404.

Then the redirect wouldn't work in the old UI eigther, because it was always hacky redirect in the Javascript. I agree it is not a best idea, but it is what it is and there was not a better way found at the time of the merge and it is in 3.1 now, so we should fix the major cases now. To make the host form js feel less hacky is too much of an ask for 3.1 cherry picks IMO :)
I'll hack on it for a bit and see if there is something that can be done.

Reply:
Argh, you're just making me find more bugs. I thought you might be able to use the name input field. However, this also doesn't work well. For example, I tried to change the hostname:
* Go to the Interfaces tab
* Change DNS name
* Click OK to close the dialog
* Go to the Host tab

Now you'll see the host no longer has an FQDN while it previously did. This is with an unmanaged host on Foreman 3.0.

However, I do think the name field itself is the most reliable one since that will actually be stored in the DB and will be searched for. So perhaps it should be gotten from the form field `host[name]`?

Reply:
I thought so as well, once opening this PR, but that is not true, the actual `host.name` is calculated by https://github.com/theforeman/foreman/blob/2152852713f100e5de8c71408fb6cb8254cddb4f/app/services/name_synchronizer.rb on a `NIC` save.
The redirect object after save is dropped (and automatically followed by jquery without any option to get that redirect object and see the current responseURL). I'm getting headache because this should BE SO SIMPLE!  and yet I can't find any good enough solution.

Reply:
It should, but as many things in life, theory and practice do not overlap :(

Just to get this whole thing clear, even if it's just in my head. The code we're talking about is:
https://github.com/theforeman/foreman/blob/2152852713f100e5de8c71408fb6cb8254cddb4f/app/assets/javascripts/host_edit.js#L161-L201

And really, in particular the success bit:

https://github.com/theforeman/foreman/blob/2152852713f100e5de8c71408fb6cb8254cddb4f/app/assets/javascripts/host_edit.js#L173-L185

That needs to ensure you're on the correct page. In particular the whole success bit. I still don't really see a reason we can't make sure `response` has a `Location` field and then push that URL to the state.

For the creation we have:
https://github.com/theforeman/foreman/blob/2152852713f100e5de8c71408fb6cb8254cddb4f/app/controllers/hosts_controller.rb#L99

Then for updating we have:
https://github.com/theforeman/foreman/blob/2152852713f100e5de8c71408fb6cb8254cddb4f/app/controllers/hosts_controller.rb#L117
This means both go through `current_host_details_path`:
https://github.com/theforeman/foreman/blob/2152852713f100e5de8c71408fb6cb8254cddb4f/app/controllers/hosts_controller.rb#L926-L928

So what's the reason we can't use `Location` header from that returned response? The server already knows the URL and sends it. We just have to use it.

Note that according to the [jquery docs](https://api.jquery.com/jQuery.ajax/) you can specify success as `Type: Function( Anything data, String textStatus, jqXHR jqXHR )`. The [jqXHR](https://api.jquery.com/jQuery.ajax/#jqXHR) argument has a `getResponseHeader(name)` method, so I think it should work to change the signature and use `new_url = jqXHR.getResponseHeader('Location')`.

Reply:
I thought it would be in the header, but apparently that object (jqXHR) has headers for the redirect, only content for the final page, I thought we could use `responseURL`, but this is not implemented in jquery: https://github.com/jquery/jquery/issues/4339

```javascript
< jqXHR.getResponseHeader('Location')
> null

< jqXHR.getAllResponseHeaders()
> `cache-control: max-age=0, private, must-revalidate\r\ncontent-security-policy: default-src 'self' localhost:3808; child-src 'self'; connect-src 'self' ws: wss: localhost:3808; font-src 'self' data: http://localhost:3808; img-src 'self' data: localhost:3808; script-src 'unsafe-eval' 'unsafe-inline' 'self' localhost:3808; style-src 'unsafe-inline' 'self' localhost:3808\r\ncontent-type: text/html; charset=utf-8\r\netag: W/"4ae9ec52826ccc3a67a7eadbd03c41f1"\r\ntransfer-encoding: chunked\r\nx-content-type-options: nosniff\r\nx-download-options: noopen\r\nx-frame-options: sameorigin\r\nx-permitted-cross-domain-policies: none\r\nx-request-id: d6000a26-bda2-44cb-be4f-93d44f9f4469\r\nx-runtime: 0.125798\r\nx-xss-protection: 1; mode=block\r\n`
```

Reply:
Yeah, it would most definitelly be better, but we would drop IE10 support and it would mean rewriting quite a significant portion of `host_edit.js` which I'd not recommend cherry-picking in 3.1

Reply:
Are you thinking about the fetch API that IE10 doesn't support? Couldn't you use XmlHttpRequest, which is available for a long time?

Also, it's unclear if we actually need to support IE10. Patternfly 4 doesn't and we've included that since Foreman 2.1: https://github.com/theforeman/theforeman.org/pull/1933. Perhaps we already broke it anyway?